presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/StaticHiveCluster.java

```diff
@@ -67,12 +82,27 @@ public StaticHiveCluster(List<URI> metastoreUris, String metastoreUsername, Hive
     public HiveMetastoreClient createMetastoreClient()
     {
         List<HostAndPort> metastores = new ArrayList<>(addresses);
-        Collections.shuffle(metastores.subList(1, metastores.size()));
+        metastores.sort(Comparator.comparingLong(m -> concurrentMap.get(m).getBackoffDelayNanos()));
 
         TException lastException = null;
         for (HostAndPort metastore : metastores) {
             try {
-                HiveMetastoreClient client = clientFactory.create(metastore);
+                HiveMetastoreClient client = new FailureAwareHiveMetaStoreClient(clientFactory.create(metastore), new ResponseHandle()
+                {
+                    @Override
+                    public void success()
+                    {
+                        concurrentMap.get(metastore).success();
+                    }
+
+                    @Override
+                    public void failed(TException e)
+                    {
+                        if (e.getCause() instanceof SocketTimeoutException) {
```

Comment:
Can `SocketTimeoutException` be an indirect cause (`e.getCause().getCause()`)?
Can there be any other interesting exception that you should care about? Metastore may fail due to other reasons as well -- like it cannot reach database,.